File: kotlin/src/test/kotlin/com/bhengubv/circleai/pets/PetsTest.kt
// PetsTest.kt
//
// Verifies the CircleAI.Pets port against the C# reference:
//   - add/get; Pets ordered by name; vaccinationsFor newest-first; weightHistory
//     oldest-first; upcomingAppointments = future (UTC now), ordered ASC
//   - domain-context constants; adapter enrichment + a domain helper

package com.bhengubv.circleai.pets

import com.bhengubv.circleai.companion.support.FakeCompanionSession
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Test
import java.time.Instant
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class PetsTest {

    private val dob = Instant.parse("2020-07-01T00:00:00Z")
    private val past = Instant.parse("2020-01-01T00:00:00Z")

    @Test
    fun `pets vaccinations weights and appointments`() {
        val b = InMemoryPetsBoard()
        b.add(Pet("p2", "Zeus", "dog", "Boxer", dob))
        b.add(Pet("p1", "Ace", "cat", null, dob))
        assertEquals(listOf("Ace", "Zeus"), b.pets.map { it.name })
        assertEquals("Zeus", b.getPet("p2")!!.name)

        b.recordVaccination(Vaccination("p1", "Rabies", past, null))
        b.recordVaccination(Vaccination("p1", "FVRCP", past.plusSeconds(86400), null)) // newer
        assertEquals(listOf("FVRCP", "Rabies"), b.vaccinationsFor("p1").map { it.vaccine }) // newest-first

        b.recordWeight(WeightSample("p1", 4.0, past.plusSeconds(100)))
        b.recordWeight(WeightSample("p1", 3.0, past)) // earlier
        assertEquals(listOf(3.0, 4.0), b.weightHistory("p1").map { it.weightKg }) // oldest-first

        val future = Instant.now().plusSeconds(86400)
        b.schedule(VetAppointment("ap1", "p1", "checkup", future, "Dr Vet"))
        b.schedule(VetAppointment("ap2", "p1", "old", past, "Dr Vet")) // in the past -> excluded
        assertEquals(listOf("ap1"), b.upcomingAppointments().map { it.apptId })
    }

    @Test
    fun `domain context and adapter`() = runTest {
        assertTrue(PetsDomainContext.SYSTEM_PROMPT_SNIPPET.startsWith("[DOMAIN: Pets]"))
        assertTrue("Vet_Referral_Required" in PetsDomainContext.complianceFlags)

        val fake = FakeCompanionSession()
        val a = PetsCompanionAdapter(fake)
        a.sendAsync("hi")
        assertTrue(fake.lastMessage!!.startsWith("[DOMAIN: Pets]"))
        a.triageSymptomAsync("dog", "Boxer", "limping")
        assertTrue(fake.lastMessage!!.contains("Triage this pet health concern"))
    }
}

File: rust/tests/pets_test.rs
//! pets_test.rs
//!
//! Ports the behaviour of `CircleAI.Pets`: pet registry (name-ordered),
//! vaccination log (newest-first), weight history (oldest-first), and upcoming
//! vet appointments (future-only, time-ordered).

use chrono::{Duration, Utc};
use circle_ai::pets::{
    IPetsBoard, InMemoryPetsBoard, Pet, Vaccination, VetAppointment, WeightSample,
};

#[test]
fn pets_added_and_name_ordered() {
    let board = InMemoryPetsBoard::new();
    assert!(board.get_pet("p1").is_none());
    board.add(Pet::new("p2", "Rex", "Dog", Some("Lab".into()), Utc::now()));
    board.add(Pet::new("p1", "Milo", "Cat", None, Utc::now()));
    let pets = board.pets();
    let names: Vec<&str> = pets.iter().map(|p| p.name.as_str()).collect();
    assert_eq!(names, vec!["Milo", "Rex"]);
}

#[test]
fn vaccinations_newest_first() {
    let board = InMemoryPetsBoard::new();
    board.record_vaccination(Vaccination::new("p1", "Rabies", Utc::now() - Duration::days(365), None));
    board.record_vaccination(Vaccination::new("p1", "Booster", Utc::now(), Some(Utc::now() + Duration::days(365))));
    board.record_vaccination(Vaccination::new("p2", "Rabies", Utc::now(), None));

    let vax = board.vaccinations_for("p1");
    let names: Vec<&str> = vax.iter().map(|v| v.vaccine.as_str()).collect();
    assert_eq!(names, vec!["Booster", "Rabies"]);
}

#[test]
fn weight_history_oldest_first() {
    let board = InMemoryPetsBoard::new();
    board.record_weight(WeightSample::new("p1", 5.0, Utc::now()));
    board.record_weight(WeightSample::new("p1", 4.0, Utc::now() - Duration::days(30)));
    board.record_weight(WeightSample::new("p2", 9.0, Utc::now()));

    let hist = board.weight_history("p1");
    let weights: Vec<f64> = hist.iter().map(|w| w.weight_kg).collect();
    assert_eq!(weights, vec![4.0, 5.0]); // oldest-first
}

#[test]
fn upcoming_appointments_future_only_time_ordered() {
    let board = InMemoryPetsBoard::new();
    let now = Utc::now();
    board.schedule(VetAppointment::new("a1", "p1", "Checkup", now + Duration::days(3), "Dr V"));
    board.schedule(VetAppointment::new("a2", "p1", "Shots", now + Duration::days(1), "Dr V"));
    board.schedule(VetAppointment::new("a3", "p2", "Past", now - Duration::days(1), "Dr V")); // past

    let up = board.upcoming_appointments();
    let ids: Vec<&str> = up.iter().map(|a| a.appt_id.as_str()).collect();
    assert_eq!(ids, vec!["a2", "a1"]);
}
